src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs

```diff
@@ -181,15 +181,22 @@ public bool TryGetReceiptsIterator(long blockNumber, Keccak blockHash, out Recei
 
             var result = CanGetReceiptsByHash(blockNumber);
             var receiptsData = _blocksDb.GetSpan(blockHash);
-            IReceiptsRecovery.IRecoveryContext? recoveryContext = null;
+
+
+            Func<IReceiptsRecovery.IRecoveryContext?> recoveryContextFactory = () => null;
 
             if (_storageDecoder.IsCompactEncoding(receiptsData))
             {
-                Block block = _blockTree.FindBlock(blockHash);
-                recoveryContext = _receiptsRecovery.CreateRecoveryContext(block!);
+                recoveryContextFactory = () =>
```

Comment:
Capture this closure in the constructor rather than allocating each time and then just assign:
```csharp
private readobly Func<IReceiptsRecovery.IRecoveryContext?> _compactFactory;

public PersistentReceiptStorage( ... )
{
    ...
    _compactFactory = () => {
        Block block = _blockTree.FindBlock(blockHash);
        return _receiptsRecovery.CreateRecoveryContext(block!);
    };
}
```

```
recoveryContextFactory = _compactFactory;
```

Reply:
It need the blockhash unfortunately.